tirocks/src/listener.rs

```diff
@@ -0,0 +1,258 @@
+    /// returns. Otherwise, RocksDB may be blocked.
+    ///
+    /// @param db a pointer to the rocksdb instance which just compacted
+    ///   a file.
+    /// @param ci a reference to a CompactionJobInfo struct. 'ci' is released
+    ///  after this function is returned, and must be copied if it is needed
+    ///  outside of this function.
+    fn on_compaction_completed(&self, _: &CompactionJobInfo) {}
+
+    /// A callback function for RocksDB which will be called each time when
+    /// a registered RocksDB uses multiple subcompactions to compact a file. The
+    /// callback is called by each subcompaction and in the same thread.
+    fn on_subcompaction_begin(&self, _: &SubcompactionJobInfo) {}
+
+    /// A callback function for RocksDB which will be called each time when
+    /// a registered RocksDB uses multiple subcompactions to compact a file. The
+    /// callback is called by each subcompaction and in the same thread.
+    fn on_subcompaction_completed(&self, _: &SubcompactionJobInfo) {}
+
+    /// A callback function for RocksDB which will be called after an external
+    /// file is ingested using IngestExternalFile.
+    ///
+    /// Note that the this function will run on the same thread as
+    /// IngestExternalFile(), if this function is blocked, IngestExternalFile()
+    /// will be blocked from finishing.
+    fn on_external_file_ingested(&self, _: &ExternalFileIngestionInfo) {}
+
+    /// A callback function for RocksDB which will be called before setting the
+    /// background error status to a non-OK value. The new background error status
+    /// is provided in `bg_error` and can be modified by the callback. E.g., a
+    /// callback can suppress errors by resetting it to Status::OK(), thus
+    /// preventing the database from entering read-only mode. We do not provide any
+    /// guarantee when failed flushes/compactions will be rescheduled if the user
+    /// suppresses an error.
+    ///
+    /// Note that this function can run on the same threads as flush, compaction,
+    /// and user writes. So, it is extremely important not to perform heavy
+    /// computations or blocking calls in this function.
+    fn on_background_error(&self, _: BackgroundErrorReason, _: &mut Status) {}
+
+    /// A callback function for RocksDB which will be called whenever a change
+    /// of superversion triggers a change of the stall conditions.
+    ///
+    /// Note that the this function must be implemented in a way such that
+    /// it should not run for an extended period of time before the function
+    /// returns.  Otherwise, RocksDB may be blocked.
+    fn on_stall_conditions_changed(&self, _: &WriteStallInfo) {}
+}
+
+extern "C" fn destructor<E: EventListener>(ctx: *mut c_void) {
+    unsafe {
+        drop(Box::from_raw(ctx as *mut E));
+    }
+}
+
+// Maybe we should reuse db instance?
+// TODO: refactor DB implement so that we can convert DBInstance to DB.
```

Comment:
I don't understand this comment. `on_flush_begin` will accept a DB after we rust-ify the DB type?

Reply:
Yes. More precisely, there will be no `crocksdb_t`, and it will accept `*mut rocksdb_DB` directly.